core/src/actor.rs

```diff
@@ -86,6 +90,94 @@ impl WinternitzDerivationPath {
     }
 }
 
+fn calc_tweaked_keypair(
+    keypair: &Keypair,
+    merkle_root: Option<TapNodeHash>,
+) -> Result<Keypair, BridgeError> {
+    Ok(keypair.add_xonly_tweak(
+        &SECP,
+        &TapTweakHash::from_key_and_tweak(keypair.x_only_public_key().0, merkle_root).to_scalar(),
+    )?)
+}
+
+fn calc_tweaked_xonly_pk(
+    pubkey: XOnlyPublicKey,
+    merkle_root: Option<TapNodeHash>,
+) -> Result<XOnlyPublicKey, BridgeError> {
+    Ok(pubkey
+        .add_tweak(
+            &SECP,
+            &TapTweakHash::from_key_and_tweak(pubkey, merkle_root).to_scalar(),
+        )?
+        .0)
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct TweakCache {
+    // A cache that holds tweaked keys so that we do not need to repeatedly calculate them.
+    // This cache will hold data for only one deposit generally because we need to clone the holder of Actor(owner or verifier)
+    // to spawned threads during deposit. (Because all grpc functions have &self, we also need to clone Actor to a mutable instance
+    // to modify the caches)
+    tweaked_key_cache: HashMap<(XOnlyPublicKey, Option<TapNodeHash>), XOnlyPublicKey>,
+    // A cache to hold actors own tweaked keys.
+    tweaked_keypair_cache: HashMap<Option<TapNodeHash>, Keypair>,
+    // Saved keypair so that user cannot accidentally call get_own_tweaked_keypair with a different keypair.
+    saved_keypair: Option<Keypair>,
+}
+
+impl TweakCache {
+    fn get_own_tweaked_keypair(
+        &mut self,
+        keypair: &Keypair,
+        merkle_root: Option<TapNodeHash>,
+    ) -> Result<&Keypair, BridgeError> {
+        if let Some(saved_keypair) = self.saved_keypair {
+            if saved_keypair != *keypair {
+                return Err(BridgeError::Error(format!(
+                    "TweakCache used with a different keypair, saved: {:?}, given: {:?}",
+                    saved_keypair, keypair
+                )));
+            }
+        } else {
+            self.saved_keypair = Some(*keypair);
```

Comment:
The first call saves the keypair, then the rest uses it? why not save it on creation?